### PR TITLE
Alpaca finetuning with LoRA 1/n

### PR DESCRIPTION
**Pull request by awaelchli**
- Adds dataloading and preprocessing for Alpaca
- Adds a basic nano-gpt-like training loop for the Alpaca instruction tuning
- Integrates LoRA (#44)

### CODE REVIEW

File: finetune.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    val_data: np.ndarray,
+) -> None:
+    """The training loop.
+
+    Loosely based on the nanoGPT implementation: https://github.com/karpathy/nanoGPT.
+    """
+
+    iter_num = 0
+
+    while True:
```
Comment (Borda) [inline finetune.py:84]:
```suggestion
    for iter_num in range(max_iters):
```

File: finetune.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+        scheduler.step()
+        optimizer.zero_grad()
+
+        dt = time.time() - t0
+        if iter_num % log_interval == 0:
+            fabric.print(f"iter {iter_num}: loss {loss.item():.4f}, time: {dt*1000:.2f}ms")
+        iter_num += 1
+
+        if iter_num > max_iters:
+            break
```
Comment (Borda) [inline finetune.py:119]:
```suggestion
```

**Review by lantiga — APPROVED** [review]
Looks great!

File: README.md [hunk 3]
```diff
@@ -115,6 +115,26 @@ See `python generate.py --help` for more options.
 
 &nbsp;
 
+## Finetune the model
+
+We provide a simple training script in `finetune.py` that finetunes a pretrained model on the Alpaca dataset and also integrates [LoRA]().
```
Comment (lantiga) [inline README.md:120]:
```suggestion
We provide a simple training script in `finetune.py` that finetunes a pretrained model on the Alpaca dataset and also integrates [LoRA](https://arxiv.org/abs/2106.09685).
```

Comment (awaelchli) [reply README.md]: updated

**Pull request merged**